File: test/scalability_3/runtest.pl
#!/usr/bin/perl
#  file     : $HeadURL$
#  revision : $LastChangedRevision$  $LastChangedBy$
#  date     : $LastChangedDate$
#  email    : jdetaeye@users.sourceforge.net

use strict;
use warnings;
use Env qw(EXECUTABLE);
use Time::HiRes qw ( gettimeofday tv_interval );

my %runtimes;

for (my $counter=500; $counter <= 2000; $counter+=500)
  {
  print "\ncounter $counter\n";
  open(OUTFILE, ">input.xml");

  # Print a header
  printf OUTFILE
    "<PLAN xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\">\n" .
    "<DESCRIPTION>Single buffer plan with $counter demands</DESCRIPTION>\n" .
    "<CURRENT>2006-01-01T00:00:00</CURRENT>\n" .
    "<COMMANDS>\n" .
    "<COMMAND xsi:type=\"COMMAND_SOLVE\" VERBOSE=\"false\">" .
    "  <SOLVER NAME=\"MRP\" xsi:type=\"SOLVER_MRP\" CONSTRAINTS=\"0\"/>" .
    "</COMMAND>\n" .
    "<COMMAND xsi:type=\"COMMAND_SAVE\" FILENAME=\"output.xml\"/>\n" .
    "</COMMANDS>\n" .
    "<ITEMS>\n" .
      "\t<ITEM NAME=\"ITEM\">" .
      "<OPERATION NAME=\"Delivery ITEM\" xsi:type=\"OPERATION_FIXED_TIME\"/>" .
      "</ITEM>\n" .
    "</ITEMS>\n" .
    "<OPERATIONS>\n" .
      "\t<OPERATION NAME=\"Make ITEM\" xsi:type=\"OPERATION_FIXED_TIME\"/>\n" .
    "</OPERATIONS>\n" .
    "<BUFFERS>\n" .
      "\t<BUFFER NAME=\"BUFFER\"><ONHAND>10</ONHAND>" .
      "<PRODUCING NAME=\"Make ITEM\"/>" .
      "</BUFFER>\n" .
    "</BUFFERS>\n" .
    "<FLOWS>\n" .
      "\t<FLOW xsi:type=\"FLOW_START\"><OPERATION NAME=\"Delivery ITEM\"/>" .
      "<BUFFER NAME=\"BUFFER\"/>" .
      "<QUANTITY>-1</QUANTITY></FLOW>\n" .
      "\t<FLOW xsi:type=\"FLOW_END\"><OPERATION NAME=\"Make ITEM\"/>" .
      "<BUFFER NAME=\"BUFFER\"/>" .
      "<QUANTITY>1</QUANTITY></FLOW>\n" .
    "</FLOWS>\n" .
    "<DEMANDS>\n";

  # A loop to print all demand
  for (my $cnt = 0; ++$cnt < $counter; )
  {
  	my $month = int(rand 12)+1;
  	my $day = int(rand 28)+1;
  	if ($month<10) {$month = "0" . $month;}
  	if ($day<10) {$day = "0" . $day;}
    printf OUTFILE
      "<DEMAND NAME=\"DEMAND $cnt\" QUANTITY=\"10\" " .
      "DUE=\"2005-${month}-${day}T00:00:00\" " .
      "PRIORITY=\"1\">" .
      "<ITEM NAME=\"ITEM\"/>" .
      "</DEMAND>\n";
  }

  # Finalize the input
  printf OUTFILE "</DEMANDS></PLAN>\n";
  close(OUTFILE);

	my $starttime = [gettimeofday];
  system("$EXECUTABLE ./input.xml");
  die "Planner exited abnormally\n" if $? > 0;

  # Measure the time
  $runtimes{$counter} = tv_interval($starttime);
  printf "time: %.3f\n", $runtimes{$counter};
  }

# Define failure criterium
if ($runtimes{2000} > $runtimes{500}*4*1.1)
{
	die "\nTest failed. Run time is not linear with model size.\n";
}

print "\nTest passed\n";
exit;

File: test/scalability_3/runtest.py
#!/usr/bin/env python3
#
# Copyright (C) 2007 by frePPLe bv
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#
import os, sys, random

runtimes = {}

for counter in [500, 1000, 1500, 2000]:
    print("\ncounter", counter)
    out = open("input.xml", "wt")

    # Print a header
    print(
        '<plan xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">\n'
        + "<description>Single buffer plan with $counter demands</description>\n"
        + "<current>2009-01-01T00:00:00</current>\n"
        + "<items>\n"
        + '\t<item name="ITEM">'
        + '<operation name="Delivery ITEM" xsi:type="operation_fixed_time"/>'
        + "</item>\n"
        + "</items>\n"
        + "<operations>\n"
        + '\t<operation name="Make ITEM" xsi:type="operation_fixed_time"/>\n'
        + "</operations>\n"
        + "<buffers>\n"
        + '\t<buffer name="BUFFER"><onhand>10</onhand>'
        + '<producing name="Make ITEM"/>'
        + "</buffer>\n"
        + "</buffers>\n"
        + "<flows>\n"
        + '\t<flow xsi:type="flow_start"><operation name="Delivery ITEM"/>'
        + '<buffer name="BUFFER"/>'
        + "<quantity>-1</quantity></flow>\n"
        + '\t<flow xsi:type="flow_end"><operation name="Make ITEM"/>'
        + '<buffer name="BUFFER"/>'
        + "<quantity>1</quantity></flow>\n"
        + "</flows>\n"
        + "<demands>",
        file=out,
    )

    # A loop to print all demand
    for cnt in range(counter):
        month = "%02d" % (int(random.uniform(0, 12)) + 1)
        day = "%02d" % (int(random.uniform(0, 28)) + 1)
        print(
            (
                '<demand name="DEMAND %d" quantity="10" '
                + 'due="2009-%s-%sT00:00:00" '
                + 'priority="1">'
                + '<item name="ITEM"/>'
                + "</demand>"
            )
            % (cnt, month, day),
            file=out,
        )

    # Finalize the input
    print(
        "</demands>\n"
        + "<?python\n"
        + "import frepple\n"
        + "frepple.solver_mrp(constraints=0).solve()\n"
        + 'frepple.saveXMLfile("output.xml")\n'
        + "?>\n"
        + "</plan>",
        file=out,
    )
    out.close()

    # Run the executable
    starttime = os.times()
    out = os.popen(os.environ["EXECUTABLE"] + "  ./input.xml")
    while True:
        i = out.readline()
        if not i:
            break
        print(i.strip())
    if out.close() != None:
        print("Planner exited abnormally")
        sys.exit(1)

    # Measure the time
    endtime = os.times()
    runtimes[counter] = endtime[4] - starttime[4]
    print("time: %.3f" % runtimes[counter])

# Define failure criterium
if runtimes[2000] > runtimes[500] * 4 * 1.2:
    print("\nTest failed. Run time is not linear with model size.")
    sys.exit(1)

print("\nTest passed")
